options/logging.go

```diff
@@ -0,0 +1,221 @@
+	if lp.IsMulti {
+		return lp.convertMultiMessage(lp.Data)
+	}
+	return lp.convertMessage(lp.Data)
+}
+
+func (lp *LoggingPayload) convertMultiMessage(value interface{}) (message.Composer, error) {
+	switch data := value.(type) {
+	case string:
+		return lp.convertMultiMessage(strings.Split(data, "\n"))
+	case []byte:
+		payload, err := splitBSONSlice(lp.Format == LoggingPayloadFormatBSON, data)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		return lp.convertMultiMessage(payload)
+	case []string:
+		batch := []message.Composer{}
+		for _, str := range data {
+			elem, err := lp.produceMessage([]byte(str))
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+			batch = append(batch, elem)
+		}
+		return message.NewGroupComposer(batch), nil
+	case [][]byte:
+		batch := []message.Composer{}
+		for _, dt := range data {
+			elem, err := lp.produceMessage(dt)
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+			batch = append(batch, elem)
+		}
+		return message.NewGroupComposer(batch), nil
+	case []interface{}:
+		batch := []message.Composer{}
+		for _, dt := range data {
+			elem, err := lp.convertMessage(dt)
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+			batch = append(batch, elem)
+		}
+		return message.NewGroupComposer(batch), nil
+	default:
+		return message.ConvertToComposer(lp.Priority, value), nil
+	}
+}
+
+func (lp *LoggingPayload) convertMessage(value interface{}) (message.Composer, error) {
+	switch data := value.(type) {
+	case string:
+		return lp.produceMessage([]byte(data))
+	case []byte:
+		return lp.produceMessage(data)
+	case []string:
+		return message.ConvertToComposer(lp.Priority, data), nil
+	case [][]byte:
+		return message.NewLineMessage(lp.Priority, byteSlicesToStringSlice(data)), nil
+	case []interface{}:
+		return message.NewLineMessage(lp.Priority, data...), nil
+	default:
+		return message.ConvertToComposer(lp.Priority, value), nil
+	}
+}
+
+func (lp *LoggingPayload) produceMessage(data []byte) (message.Composer, error) {
+	switch lp.Format {
+	case LoggingPayloadFormatJSON:
+		payload := message.Fields{}
+		if err := json.Unmarshal(data, &payload); err != nil {
+			return nil, errors.Wrap(err, "problem parsing json from message body")
+		}
+
+		if lp.AddMetadata {
+			return message.NewFields(lp.Priority, payload), nil
+		}
+
+		return message.NewSimpleFields(lp.Priority, payload), nil
+	case LoggingPayloadFormatBSON:
+		payload := message.Fields{}
+		if err := bson.Unmarshal(data, &payload); err != nil {
+			return nil, errors.Wrap(err, "problem parsing bson from message body")
+		}
+
+		if lp.AddMetadata {
+			return message.NewFields(lp.Priority, payload), nil
+		}
+
+		return message.NewSimpleFields(lp.Priority, payload), nil
+	default: // includes string case.
+		if lp.AddMetadata {
+			return message.NewSimpleBytesMessage(lp.Priority, data), nil
+		}
+
+		return message.NewBytesMessage(lp.Priority, data), nil
+	}
+}
+
```

Comment:
Rather than pass a bool that immediately returns for a non-bson case, the use of splitBSONSlice on L100 should be split into two different functions since they have entirely different behavior and allowing non-bson to be passed to a function that has the word "BSON" in it sounds deceptive.